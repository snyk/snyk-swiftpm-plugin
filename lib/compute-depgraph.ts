import { DepGraph, DepGraphBuilder } from '@snyk/dep-graph';
import { execute } from './subprocess';
import * as path from 'path';

export type DepTreeNode = {
  name: string;
  url: string;
  version: string;
  path: string;
  dependencies: DepTreeNode[];
};

export async function computeDepGraph(
  root: string,
  targetFile: string,
  additionalArgs?: string[],
): Promise<DepGraph> {
  const fs = require('fs');
  const os = require('os');

  //create temporary folder
  const tempDir = os.tmpdir();

  //move the .build folder to temporary folder
  fs.renameSync('.build', `${tempDir}/.build`);

  //delete temporary foler
  fs.rmdirSync(tempDir, { recursive: true });

  try {
    const defaultArgs = [
      'package',
      '--package-path',
      path.dirname(targetFile),
      'show-dependencies',
      '--format',
      'json',
    ];

    const args = additionalArgs
      ? defaultArgs.concat(additionalArgs)
      : defaultArgs;
    const result = await execute('swift', args, { cwd: root });
    const depTree: DepTreeNode = JSON.parse(result);
    const depGraph: DepGraph = convertToGraph(depTree);

    return depGraph;
  } catch (e) {
    throw new Error('Unable to generate dependency tree');
  }
}

function traverseTree(
  rootNode: DepTreeNode,
  builder: DepGraphBuilder,
  rootNodeId?: string,
) {
  const childNodes = rootNode.dependencies;

  childNodes?.forEach((node) => {
    const { url, version } = node;
    const name = url
      .replace(/https:\/\//g, '')
      .replace(/http:\/\//g, '')
      .replace(/.git/g, '');
    const parentName =
      rootNodeId ||
      rootNode.url
        .replace(/https:\/\//g, '')
        .replace(/http:\/\//g, '')
        .replace(/.git/g, '');

    const nodeId = `${name}@${version}`;
    const parentNodeId = `${parentName}@${rootNode.version}`;

    builder.addPkgNode({ name, version }, nodeId);
    builder.connectDep(rootNodeId || parentNodeId, nodeId);

    traverseTree(node, builder);
  });
}

function convertToGraph(rootNode: DepTreeNode): DepGraph {
  const { name, version } = rootNode;
  const depGraphBuilder = new DepGraphBuilder(
    { name: 'swift' },
    { name, version },
  );

  traverseTree(rootNode, depGraphBuilder, depGraphBuilder.rootNodeId);

  return depGraphBuilder.build();
}
