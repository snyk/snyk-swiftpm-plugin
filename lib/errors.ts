export class SwiftError extends Error {
    constructor(message: string, swiftErrStack: any) {
        super(message);
        this.stack = swiftErrStack;
    }
}
