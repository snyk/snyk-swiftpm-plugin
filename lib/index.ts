import { computeDepGraph as swiftDepGraph } from './swiftpm/compute-depgraph';
import { computeDepGraph as carthageDepGraph } from './carthage/deps';
import { lookpath } from 'lookpath';
import * as path from 'path';

interface Options {
  debug?: boolean;
  file?: string;
  args?: string[];
}

// we assume that swift considers folders as packages instead of manifest files
function pathToPosix(fpath) {
  const parts = fpath.split(path.sep);
  parts.pop();
  if (parts.length === 0) {
    return './';
  }
  return parts.join(path.posix.sep);
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export async function inspect(
  root: string,
  targetFile: string,
  options?: Options,
) {
  const filename = path.basename(targetFile);
  const supportedPackageManagers = ['Swift Package Manager', 'Carthage'];
  const supportedTargetFiles = {
    'Package.swift': swiftDepGraph,
    'Cartfile.resolved': carthageDepGraph,
  };
  if (!(filename in supportedTargetFiles)) {
    throw new Error(
      `${filename} is not supported by ${supportedPackageManagers.join(' or ')}. ` +
        `Please provide with path to ${Object.keys(supportedTargetFiles).join(' or ')} files.`,
    );
  }
  if (filename === 'Package.swift') {
    const swiftPath = await lookpath('swift');
    if (!swiftPath) {
      throw new Error(
        'The "swift" command is not available on your system. ' +
          'To scan your dependencies in the CLI, you must ensure you have ' +
          'first installed the relevant package manager.',
      );
    }
  }

  const computeDepGraph = supportedTargetFiles[filename];
  const depGraph = await computeDepGraph(root, targetFile, options?.args);
  if (!depGraph) {
    throw new Error(`Failed to scan ${targetFile}`);
  }
  return {
    plugin: {
      name: 'snyk-swiftpm-plugin',
      runtime: 'unknown',
      targetFile: `${pathToPosix(targetFile)}${filename}`,
    },
    dependencyGraph: depGraph,
  };
}
