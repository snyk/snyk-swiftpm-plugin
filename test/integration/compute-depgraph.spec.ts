import { computeDepGraph } from '../../lib/compute-depgraph';
import * as path from 'path';

describe('compute-depgraph', () => {
  // Run locally for verifications. Or even better, add OSX runner to CircleCI and run it there as well.
  it.skip('should successfully create snyk dep graph from swift-pm dep tree', async () => {
    const targetFile = path.join(__dirname, '../fixtures/Package.swift');
    const result = await computeDepGraph(
      path.join(__dirname, '../fixtures'),
      targetFile,
    );

    expect(result).toMatchSnapshot();
  }, 100000);
});
