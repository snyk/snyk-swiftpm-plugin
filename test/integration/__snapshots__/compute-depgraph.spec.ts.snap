// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compute-depgraph should successfully create snyk dep graph from swift-pm dep tree 1`] = `
Object {
  "graph": Object {
    "nodes": Array [
      Object {
        "deps": Array [
          Object {
            "nodeId": "github.com/grpc/grpc-swift@1.16.0",
          },
        ],
        "nodeId": "root-node",
        "pkgId": "swiftPM-spike@unspecified",
      },
      Object {
        "deps": Array [
          Object {
            "nodeId": "github.com/apple/swift-nio@2.53.0",
          },
          Object {
            "nodeId": "github.com/apple/swift-nio-http2@1.26.0",
          },
          Object {
            "nodeId": "github.com/apple/swift-nio-transport-services@1.17.0",
          },
          Object {
            "nodeId": "github.com/apple/swift-nio-extras@1.19.0",
          },
          Object {
            "nodeId": "github.com/apple/swift-protobuf@1.21.0",
          },
          Object {
            "nodeId": "github.com/apple/swift-log@1.5.2",
          },
          Object {
            "nodeId": "github.com/apple/swift-nio-ssl@2.24.0",
          },
        ],
        "nodeId": "github.com/grpc/grpc-swift@1.16.0",
        "pkgId": "github.com/grpc/grpc-swift@1.16.0",
      },
      Object {
        "deps": Array [
          Object {
            "nodeId": "github.com/apple/swift-atomics@1.1.0",
          },
          Object {
            "nodeId": "github.com/apple/swift-collections@1.0.4",
          },
        ],
        "nodeId": "github.com/apple/swift-nio@2.53.0",
        "pkgId": "github.com/apple/swift-nio@2.53.0",
      },
      Object {
        "deps": Array [],
        "nodeId": "github.com/apple/swift-atomics@1.1.0",
        "pkgId": "github.com/apple/swift-atomics@1.1.0",
      },
      Object {
        "deps": Array [],
        "nodeId": "github.com/apple/swift-collections@1.0.4",
        "pkgId": "github.com/apple/swift-collections@1.0.4",
      },
      Object {
        "deps": Array [
          Object {
            "nodeId": "github.com/apple/swift-nio@2.53.0",
          },
          Object {
            "nodeId": "github.com/apple/swift-atomics@1.1.0",
          },
        ],
        "nodeId": "github.com/apple/swift-nio-http2@1.26.0",
        "pkgId": "github.com/apple/swift-nio-http2@1.26.0",
      },
      Object {
        "deps": Array [
          Object {
            "nodeId": "github.com/apple/swift-nio@2.53.0",
          },
          Object {
            "nodeId": "github.com/apple/swift-atomics@1.1.0",
          },
        ],
        "nodeId": "github.com/apple/swift-nio-transport-services@1.17.0",
        "pkgId": "github.com/apple/swift-nio-transport-services@1.17.0",
      },
      Object {
        "deps": Array [
          Object {
            "nodeId": "github.com/apple/swift-nio@2.53.0",
          },
        ],
        "nodeId": "github.com/apple/swift-nio-extras@1.19.0",
        "pkgId": "github.com/apple/swift-nio-extras@1.19.0",
      },
      Object {
        "deps": Array [],
        "nodeId": "github.com/apple/swift-protobuf@1.21.0",
        "pkgId": "github.com/apple/swift-protobuf@1.21.0",
      },
      Object {
        "deps": Array [],
        "nodeId": "github.com/apple/swift-log@1.5.2",
        "pkgId": "github.com/apple/swift-log@1.5.2",
      },
      Object {
        "deps": Array [
          Object {
            "nodeId": "github.com/apple/swift-nio@2.53.0",
          },
        ],
        "nodeId": "github.com/apple/swift-nio-ssl@2.24.0",
        "pkgId": "github.com/apple/swift-nio-ssl@2.24.0",
      },
    ],
    "rootNodeId": "root-node",
  },
  "pkgManager": Object {
    "name": "swift",
  },
  "pkgs": Array [
    Object {
      "id": "swiftPM-spike@unspecified",
      "info": Object {
        "name": "swiftPM-spike",
        "version": "unspecified",
      },
    },
    Object {
      "id": "github.com/grpc/grpc-swift@1.16.0",
      "info": Object {
        "name": "github.com/grpc/grpc-swift",
        "version": "1.16.0",
      },
    },
    Object {
      "id": "github.com/apple/swift-nio@2.53.0",
      "info": Object {
        "name": "github.com/apple/swift-nio",
        "version": "2.53.0",
      },
    },
    Object {
      "id": "github.com/apple/swift-atomics@1.1.0",
      "info": Object {
        "name": "github.com/apple/swift-atomics",
        "version": "1.1.0",
      },
    },
    Object {
      "id": "github.com/apple/swift-collections@1.0.4",
      "info": Object {
        "name": "github.com/apple/swift-collections",
        "version": "1.0.4",
      },
    },
    Object {
      "id": "github.com/apple/swift-nio-http2@1.26.0",
      "info": Object {
        "name": "github.com/apple/swift-nio-http2",
        "version": "1.26.0",
      },
    },
    Object {
      "id": "github.com/apple/swift-nio-transport-services@1.17.0",
      "info": Object {
        "name": "github.com/apple/swift-nio-transport-services",
        "version": "1.17.0",
      },
    },
    Object {
      "id": "github.com/apple/swift-nio-extras@1.19.0",
      "info": Object {
        "name": "github.com/apple/swift-nio-extras",
        "version": "1.19.0",
      },
    },
    Object {
      "id": "github.com/apple/swift-protobuf@1.21.0",
      "info": Object {
        "name": "github.com/apple/swift-protobuf",
        "version": "1.21.0",
      },
    },
    Object {
      "id": "github.com/apple/swift-log@1.5.2",
      "info": Object {
        "name": "github.com/apple/swift-log",
        "version": "1.5.2",
      },
    },
    Object {
      "id": "github.com/apple/swift-nio-ssl@2.24.0",
      "info": Object {
        "name": "github.com/apple/swift-nio-ssl",
        "version": "2.24.0",
      },
    },
  ],
  "schemaVersion": "1.3.0",
}
`;
